generate: Adjust to Spec.Linux being a pointer

wking opened the pull request:
Catch up with the in-flight opencontainers/runtime-spec@6323157 (specs-go/config: Make Linux and Solaris omitempty (again), 2016-06-17, opencontainers/runtime-spec#502).

The “does the marshaled JSON look like '{}'?” check is a pretty cheap trick, but it was the easiest way I could think of for “is there anything useful in here?”.  An alternative approach that conditionally added an `&rspec.Linux{}` only if needed seemed too tedious.

This should not land before opencontainers/runtime-spec#502, and it may not even be worth reviewing until then.  But I'd worked up the changes while testing the runtime-spec PR, so I thought I might as well push them in case other folks wanted to see how I'd tested it.

liangchenye commented:
Your commit to runtime spec is checked in.

I don't quite like the '{}' way, how about using 'reflect.DeepEqual', like this:
```
if reflect.DeepEqual(Spec.Linux, &rspec.Linux{}) {
     spec.Linux = nil
}
```

wking commented:
On Tue, Jun 28, 2016 at 02:57:30AM -0700, 梁辰晔 (Liang Chenye) wrote:
> Your commit to runtime spec is checked in.

I'd rather hold off on landing this PR until runtime-spec cuts another
tag (#108).

> I don't quite like the '{}' way, how about using
> 'reflect.DeepEqual', like this:
> ```
> if reflect.DeepEqual(Spec.Linux, &rspec.Linux{}) {
>      spec.Linux = nil
> }
> ```

A agree that the JSON marshaling feels awkward, but it is checking
what we want (“will this serialize to really boring JSON?”).  I tried
the DeepEqual way and still had a Linux entry with:

  $ ./ocitools generate -template <(echo '{}')

My guess is that DeepEquals doesn't work because spec.Annotations is
an empty map [1] (instead of nil), or some such.

[1]: https://github.com/opencontainers/ocitools/blob/aa9cba2bdf52cddd060bf9fbe4bee47e5bfc802a/generate.go#L194

wking commented:
Rebased around #131 with 965e585 → 1cd5062.  Now that there are more
potential entry points, I needed more “Is Linux nil?” guards, but I
think I added them in the right places ;).

wking commented:
Rebased onto master with 1cd5062 → f2107e5.  Between #140 and #144,
the only thing left here is the “don't serialize Spec.Linux as an
empty JSON object” check.  Although to notice now, you have to do
something like:

  $ ocitools generate --template <(echo '{"linux": {"namespaces": [{"type": "user"}]}}') --user=host

which generates a ‘"linux": {}’ entry on master but not with this
branch.

Mashimiao commented on generate/generate.go:
```diff
@@ -959,6 +969,9 @@ func (g *Generator) SetupPrivileged(privileged bool) {
+		if g.spec.Linux == nil {
+			g.spec.Linux = &rspec.Linux{}
+		}
 		g.spec.Linux.Seccomp = nil
```
I think this is not needed any more. Because we have g.initSpecLinux()

wking commented:
On Fri, Sep 23, 2016 at 01:31:59AM -0700, Ma Shimiao wrote:
> > @@ -959,6 +969,9 @@ func (g *Generator) SetupPrivileged(privileged bool) {
>  		g.spec.Process.Capabilities = finalCapList
>  		g.spec.Process.SelinuxLabel = ""
>  		g.spec.Process.ApparmorProfile = ""
> +		if g.spec.Linux == nil {
> +			g.spec.Linux = &rspec.Linux{}
> +		}
>  		g.spec.Linux.Seccomp = nil
> 
> I think this is not needed any more. Because we have g.initSpecLinux() 

Dropped and rebased onto master with f2107e5 → 7fcc366.

Mashimiao commented:
need rebase @wking

wking commented:
On Sun, Dec 18, 2016 at 05:57:26PM -0800, Ma Shimiao wrote:
> need rebase @wking 

Rebased with 7fcc366 → 9004f6d.

Mashimiao commented:
@liangchenye  any comments?

liangchenye commented:
https://golang.org/src/encoding/json/encode.go?s=5895:5938#L585
I think this solution is the simplest way (to my knowledge) and won't bring any side-effect.

Mashimiao commented:
LGTM
